### DISCUSSION TITLE
🌏エッジランタイムとは？

### DISCUSSION THREAD

**Discussion by kazzyfrog** [lgtm-factory/lgtm-factory#82] [Scrap]
全員に共有しておいた方が良さそうなこと & 自分もまだ理解が浅いこと
なので、つらつら記録していきますー👍

---
TL;DR
- エッジランタイムとは、CDNロケーションに配置されたコンピューティングリソースおよびその実行環境
- 長所として、パフォーマンス向上や、開発者体験の向上が挙げられる
- 注意点として、Node.jsとは異なる実行環境なので、ネイティブのNode.js APIはサポートされていない

サクッとエッジランタイムについて知りたい場合は、以下のスライドを参照ください！
わかりやすいです！
https://speakerdeck.com/aiji42/biginaxiang-ke-etuzirantaimunosusume-etuzirantaimuwoyi-shi-sitakai-fa-wohazimeyou

**Comment by kazzyfrog** [comment]
先日、機能実装のデモで色々いじっていたら、fs モジュール関連でエラーが出た。

**Comment by kazzyfrog** [comment]
## エラー：Module not found: Can't resolve 'fs' - NextJS

**Reply by kazzyfrog** [reply]
ちなみに、
fsは、ファイルシステムの略。
> node:fsモジュールは、Node.jsでファイルの読み書きを行うための基本的な関数を提供するモジュールです。
参考：https://jsprimer.net/use-case/nodecli/read-file/

つまり、いろんなファイルを読み込む際に役立つ！
ref: 、[Node.js fsのメモ](https://zenn.dev/ch0c0late/articles/284e2e25a0e609)

要はこれで、fs.readdir を使って、
特定のディレクトリ配下のファイル一覧を表示できるので、やろうとした感じ。。！
そしたらエラー！

**Reply by kazzyfrog** [reply]
## 結論

Node.jsのfsモジュールなので、
そのままでは、Node.js以外の環境で動かないというだけ！

（まさに同様のスレが、Next.js のGitHub上にありました）
https://github.com/vercel/next.js/discussions/54176

**Reply by kazzyfrog** [reply]
## 解説
https://gotohayato.com/content/553/
上の記事によると、
> いろいろな原因がありえますが、 Next.js でよくあるのは クライアントサイド JS のコンテキストで fs モジュールが利用されてしまっていること です。 fs モジュールは Node.js のモジュールであり、クライアントサイド JS には含まれていません。 その fs モジュールをクライアントサイドで利用しまっているために Can't resolve 'fs' のエラーが出ます。

つまり、クライアントサイドのJSは、
Node.jsではなく、ブラウザで実行されるので、そのままではエラーになる！

また、クライアントバンドルを構築するには、
次のコンテンツを含むnext.config.jsファイルを作成する必要があるかもしれません。
```js
module.exports = {
  webpack5: true,
  webpack: (config) => {
    config.resolve.fallback = { fs: false };

    return config;
  },
};
```
参考：
https://techblg.app/articles/how-to-handle-module-not-found-cant-resolve-fs-nextjs/
https://stackoverflow.com/questions/64926174/module-not-found-cant-resolve-fs-in-next-js-application

**Reply by kazzyfrog** [reply]
同様に、エッジランタイム上も、Node.js環境ではないので、
fsを使うとエラーとなる！

**Reply by kazzyfrog** [reply]
エラーに関しては、下記で記事化しました🫠
https://zenn.dev/kazzyfrog/articles/error-fs-using-nextjs

**Comment by kazzyfrog** [comment]
## エッジランタイムとは？

参考：
next/runtimes: 
https://nextjs.org/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes
Edge Runtimeで遊んでみる
https://zenn.dev/yusukebe/articles/f3b130f69287fc

**Reply by kazzyfrog** [reply]
## [next公式のエッジランタイム](https://nextjs.org/docs/app/api-reference/edge)のページを読む

そもそも、エッジランタイムは、Node.js環境とは違い、、
- すべてのNode.js APIをサポートしているわけではありません。一部のパッケージは機能しません。
- また、増分静的再生(ISR)をサポートしていません。

ページ内にちゃんと書いてありました。。🫠🫠
> [サポートされていないAPI](https://nextjs.org/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes#unsupported-apis)
エッジランタイムには、次のようないくつかの制限があります。
> - ネイティブのNode.js APIはサポートされていません。たとえば、ファイルシステムを読み書きすることはできません。

**Reply by kazzyfrog** [reply]
https://zenn.dev/takewell/articles/136882ac35afce
ちなみに、今回のデプロイ先はCloudflare Pages。

これは、サーバーサイドの処理を、Cloudflare Workersで実行する！
これは、エッジランタイム環境！

そういえば、デプロイの実装する際にも、自分でこんなこと書いた！

>  将来的に、APIを実装する際は、エッジランタイムを使用する必要があります。
> （Cloudflareでは、現在はNext.jsの「Edge」ランタイムしか使用できません。）
> > 4.すべてのサーバーレンダリングルートがエッジランタイムを使用していることを確認してください
> [参考](https://developers.cloudflare.com/pages/framework-guides/nextjs/ssr/get-started/#4-ensure-all-server-rendered-routes-use-the-edge-runtime)
> 
> このため、サーバーサイドでの処理を実装する際には、そのファイルにて、
> 以下のような設定をする必要があることに注意してください！
> ```ts
> export const runtime = "edge";
> ```

**Comment by kazzyfrog** [comment]
## AI（gemini）に聞いた
---
## Next.jsのappディレクトリとEdge Runtimeについて

Next.jsのappディレクトリで開発を行う際に、避けて通れないのが**Edge Runtime**という概念です。

### Edge Runtimeとは？

Edge Runtimeは、Next.jsの新しいルーティングモデルであるApp Routerで導入された、**Web標準のAPIを使ったエッジ環境で動くアプリケーションの開発やテストを助けるツールキット**です。

より具体的に言うと、Edge Runtimeは以下のような特徴を持っています。

* **エッジ環境のシミュレーション:** Node.jsとは異なるエッジ環境の特性をシミュレートすることで、実際のエッジ環境で動作するアプリケーションを開発段階からテストできます。
* **Web標準APIの利用:** Fetch APIなど、Web標準のAPIを直接利用できるため、サーバーサイドレンダリング(SSR)とクライアントサイドレンダリング(CSR)の両方を統一的なコードで記述できます。
* **高性能:** エッジ環境は、ユーザーに最も近い場所に配置されているため、高速なレスポンスを実現できます。

### なぜEdge Runtimeが重要なのか？

* **パフォーマンス向上:** エッジ環境は、ユーザーに最も近い場所に配置されているため、レイテンシーを削減し、Webアプリケーションの高速化に貢献します。
* **開発効率の向上:** Web標準APIを利用できるため、学習コストが低く、開発効率が向上します。
* **柔軟性:** SSRとCSRを統一的なコードで記述できるため、アプリケーションの構造を柔軟に設計できます。

### Edge Runtimeを使うメリット

* **静的サイト生成(SSG)と動的サーバーサイドレンダリング(SSR)の統合:** Edge Runtimeは、SSGとSSRをシームレスに統合することができます。これにより、パフォーマンスと柔軟性の両方を最大限に引き出すことができます。
* **リアルタイムなデータ更新:** WebSocketsやServer-Sent Eventsなどの技術を利用することで、リアルタイムなデータ更新を実現できます。
* **グローバルな展開:** エッジネットワークを利用することで、世界中のユーザーに高速なサービスを提供できます。

### Edge Runtimeの利用例

* **リアルタイムチャットアプリケーション:** WebSocketsを利用して、リアルタイムなメッセージの送受信を実現できます。
* **パーソナライズされたコンテンツ配信:** ユーザーの行動履歴に基づいて、パーソナライズされたコンテンツを配信できます。
* **A/Bテスト:** 複数のバージョンのコンテンツを配信し、ユーザーの反応を比較することで、最適なコンテンツを特定できます。

### まとめ

Edge Runtimeは、Next.jsの開発において、より高速で柔軟なWebアプリケーションを構築するための重要なツールです。Web標準APIを利用できるため、学習コストが低く、開発効率が向上します。また、エッジ環境の特性を活かすことで、高性能なWebアプリケーションを実現できます。

**Reply by kazzyfrog** [reply]
注意すべきポイントは？
---
## Edge Runtimeで注意すべき点とfsモジュールの使用について

Edge Runtimeは、Next.jsのApp Routerで導入された強力な機能ですが、従来のサーバーサイド開発とは異なる点があるため、注意すべき点がいくつかあります。

### Edge Runtimeで注意すべき点

* **サーバーレス環境:** Edge Runtimeはサーバーレス環境で動作するため、従来のNode.js環境で自由にできたファイルシステムへのアクセスやグローバルな状態の管理などが制限されます。
* **Web標準APIの利用:** Node.jsの組み込みモジュールではなく、Fetch APIなどのWeb標準APIを主に使用します。
* **実行環境:** Edge Runtimeは、Node.jsとは異なる実行環境であり、一部のNode.jsのモジュールは直接使用できません。
* **ビルド時処理:** Edge Runtimeは、ビルド時にコードが最適化されるため、動的なコード生成や評価には注意が必要です。

### fsモジュールの使用

**Edge Runtimeでは、fsモジュールは直接使用できません。**

fsモジュールは、ファイルシステムへのアクセスを行うためのNode.jsの組み込みモジュールですが、Edge Runtimeはサーバーレス環境であり、ファイルシステムへの直接的なアクセスが制限されているためです。

### 代替手段

fsモジュールが使えない代わりに、以下の方法が考えられます。

* **データストアの利用:** データの保存には、データベースやクラウドストレージサービスを利用します。
* **サーバーサイド関数:** 必要に応じて、サーバーサイド関数を作成し、ファイルシステムへのアクセスを限定的に行うことができます。
* **ビルド時の処理:** 静的なファイルは、ビルド時に生成し、配信します。

### まとめ

Edge Runtimeは、Webアプリケーションのパフォーマンス向上に大きく貢献する一方で、従来のサーバーサイド開発とは異なる点があるため、注意が必要です。特に、fsモジュールのようなファイルシステムへのアクセスを伴う操作は、Edge Runtimeでは直接行えないことを理解しておくことが重要です。

**Comment by kazzyfrog** [comment]
## まとめ

**エッジランタイム**とは、Webアプリケーションをより高速かつ効率的に開発するためのツールキットです。
Node.js環境以外の、サーバーサイドの実行環境とも言えます。

**主な特徴とメリット**

* **高速化:** ユーザーに近いエッジサーバーで処理を行うため、レスポンス速度が向上します。
* **柔軟性:** Web標準APIを利用できるため、様々な種類のWebアプリケーションに対応できます。
* **開発効率:** サーバーレス環境のため、インフラの管理が不要になり、開発に集中できます。

**注意すべき点**

* **ファイルシステムへのアクセス制限:** Node.jsのfsモジュールなど、ファイルシステムに直接アクセスする機能は制限されています。
* **実行環境:** Node.jsとは異なる実行環境のため、一部のモジュールは使用できません。

**まとめ**

エッジランタイムは、Webアプリケーションの開発において、より高速で柔軟な開発を可能にする技術です。ただし、従来のサーバーサイド開発とは異なる点があるため、その特徴を理解した上で開発を進めることが重要です。

**Comment by kagomen** [comment]
ローカルフォントを使ったデモページ実装してる際に、自分もハマりました💧
その際こちらが参考になったので、共有！

https://vercel.com/docs/functions/og-image-generation#runtime-support


> [Runtime support](https://vercel.com/docs/functions/og-image-generation#runtime-support)
Vercel OG image generation is supported on both the [Edge runtime](https://vercel.com/docs/functions/edge-functions/edge-runtime) and [Node.js runtime](https://vercel.com/docs/functions/runtimes/node-js).
> 
> [Edge runtime](https://vercel.com/docs/functions/og-image-generation#edge-runtime)
In the Edge runtime, resources like fonts and images are loaded using the fetch method. This is effective for accessing local resources using the import.meta syntax. See [using a local image](https://vercel.com/docs/functions/edge-functions/og-image-generation/og-image-examples#using-a-local-image) for more information.
> 
> [Node.js runtime](https://vercel.com/docs/functions/og-image-generation#node.js-runtime)
For the Node.js runtime, local resources can be loaded directly using fs.readFile. Alternatively, fetch can be used to load remote resources.

**Reply by kazzyfrog** [reply]
ありがとうございます！

環境の違い、最初に押さえておかないと注意ですね！👍

**Comment by kagomen** [comment]
## エッジランタイムにおけるラウンドトリップの弊害を教えて

エッジランタイムにおけるラウンドトリップの弊害には、以下のようなものがあります：

1. レイテンシの増加:
エッジデバイスとクラウドの間で頻繁なラウンドトリップが発生すると、全体的なレイテンシが増加します。これは特にリアルタイム処理が必要なアプリケーションで問題となります。

2. 帯域幅の消費:
頻繁なデータ転送は、ネットワーク帯域幅を消費し、他のアプリケーションやサービスのパフォーマンスに影響を与える可能性があります。

3. コストの増加:
クラウドへのデータ転送やAPI呼び出しが増えると、クラウドサービスの利用コストが上昇する可能性があります。

4. 信頼性の低下:
ネットワーク接続に依存することで、接続の問題やサービスの中断が発生した場合にアプリケーションの信頼性が低下します。

5. エネルギー消費の増加:
頻繁な通信はデバイスのバッテリー消費を増加させ、特にモバイルデバイスやIoTデバイスの運用時間に影響を与えます。

6. セキュリティリスクの増加:
データが頻繁にネットワーク上を移動することで、潜在的なセキュリティリスクが増加します。

7. スケーラビリティの制限:
中央のクラウドサーバーへの依存度が高いと、大規模なデプロイメントでボトルネックが発生する可能性があります。

8. 処理の遅延:
データの往復に時間がかかるため、即時の意思決定や処理が必要な場合に問題となります。

これらの弊害を軽減するために、エッジコンピューティングでは可能な限りデータをローカルで処理し、必要最小限のデータのみをクラウドに送信するアプローチが推奨されます。

**Reply by kagomen** [reply]
## 以下の文を初心者にもわかるよう説明して

エッジがマルチロケーションなのに対して、基本的にDB
はシングルロケーションである。そうなると着弾するエッジによってはラウンドトリップが無視できなくなる可能性がある

--- 

はい、この文をより簡単に説明しましょう。

1. エッジについて：
エッジとは、ユーザーに近い場所に配置されるサーバーやデバイスのことです。「マルチロケーション」とは、これらのエッジが世界中の多くの場所に存在することを意味します。

2. DBについて：
DBはデータベースの略で、データを保存する中央の場所です。「シングルロケーション」とは、通常、このデータベースが一箇所にあることを意味します。

3. ラウンドトリップとは：
データがエッジからデータベースに行って戻ってくるまでの往復の旅のことです。

4. 問題点：
エッジは世界中にありますが、データベースは一箇所にあります。ユーザーが使用するエッジがデータベースから遠い場所にある場合、データの往復に時間がかかります。

5. 「着弾するエッジ」とは：
ユーザーのリクエストが最初に到達するエッジサーバーのことです。

6. 結論：
ユーザーのリクエストが遠くのエッジに到達した場合、データベースとの通信に時間がかかり、全体的な応答時間が遅くなる可能性があります。これが「ラウンドトリップが無視できなくなる」ということです。

簡単に言えば、エッジとデータベースの位置関係によっては、データのやり取りに時間がかかり、サービスの速度に影響を与える可能性があるということです。

**Reply by kagomen** [reply]
## 例えばラウンドトリップの大きさが原因で、エッジランタイムの制限にひっかかることがある？

はい、その通りです。エッジランタイムの制限にひっかかる可能性は十分にあります。これについて、もう少し詳しく説明しましょう。

1. エッジランタイムの制限:
多くのエッジコンピューティングプラットフォームには、実行時間の制限があります。例えば、AWS LambdaやCloudflare Workersなどのサービスでは、関数の実行時間に上限が設けられています。

2. ラウンドトリップの影響:
エッジからデータベースへのラウンドトリップに時間がかかると、その分だけ全体の処理時間が長くなります。

3. 制限超過の可能性:
ラウンドトリップの時間が長くなりすぎると、エッジランタイムの制限時間を超えてしまう可能性があります。

例を挙げて説明します：

- 仮に、エッジランタイムの制限が5秒だとします。
- エッジでの処理に2秒かかるとします。
- 通常のデータベースへのラウンドトリップが1秒で済むケースでは、合計3秒で処理が完了し、制限内に収まります。
- しかし、エッジとデータベースの距離が遠い場合、ラウンドトリップに4秒かかってしまうと、合計で6秒になります。
- この場合、5秒の制限を超えてしまい、エラーや処理の中断が発生する可能性があります。

これらの問題を回避するためには：

1. データベースのレプリケーションや分散配置
2. エッジでのキャッシュの活用
3. 非同期処理の導入
4. クリティカルなデータのみをエッジで処理し、詳細なデータ処理はバックグラウンドジョブとして実行

などの戦略が考えられます。これらの方法を用いて、ラウンドトリップの影響を最小限に抑え、エッジランタイムの制限内で処理を完了させることが重要です。

**Reply by kagomen** [reply]
## ちなみにWorkersの制限時間は何秒なのでしょうか

Cloudflare Workersの実行時間制限について説明します：

1. 基本的な制限時間:
Cloudflare Workersの標準的な実行時間制限は30秒です。これは、ほとんどのWorkersスクリプトに適用される制限です。

2. サブリクエストの制限:
Workersスクリプト内で行われるサブリクエスト（fetch()呼び出しなど）の合計時間も、この30秒の制限に含まれます。

3. Paid Planの場合:
Cloudflare Workers Paid Planを使用している場合、一部のWorkersで最大60秒まで実行時間を延長できます。

4. Durable Objectsの場合:
Durable Objectsを使用する場合、単一のリクエストに対する実行時間制限は30秒ですが、Durable Object自体は長時間（数時間または数日）存続可能です。

5. Cron Triggersの場合:
Cron Triggersを使用してスケジュールされたWorkersの場合、最大実行時間は15分です。

6. 注意点:
- これらの制限は、Cloudflareのポリシーや製品の更新により変更される可能性があります。
- 実際の使用時には、最新のドキュメントを確認することをお勧めします。

これらの制限時間を考慮すると、データベースへのラウンドトリップなどの時間のかかる処理を行う場合は、適切な設計と最適化が必要になります。例えば、非同期処理や分散キャッシュの利用、あるいはDurable Objectsの活用などが考えられます。
